text/nodejs/0014-yarn-berry.md

```diff
@@ -0,0 +1,249 @@
+# Support for Yarn "Berry" 
+
+## Summary
+
+The Paketo Node.js Buildpack supports building apps which use
+[Yarn](https://yarnpkg.com/) as a package/project management tool. Over the
+last few years, the developers of Yarn have released new major versions (v2 &
+v3), collectively called Yarn "Berry", which represent a completely new
+direction for managing dependencies in Node.js projects. Chief among the
+changes is the introduction of [Plug 'n
+Play](https://yarnpkg.com/features/pnp), a system which removes the reliance on
+`node_modules` and can eliminate the need to run `yarn install` at all. The
+Yarn implementation buildpacks do not currently support building apps which use
+Berry.  
+
+There have been quite a number of changes in the behaviour of Yarn with the
+advent of Berry. Along with performance improvements and optimizations, many
+CLI operations and methods of configuration have been deprecated or removed,
+including those crucial to the functioning of the buildpack as it exists today. 
+
+## Proposal
+
+Yarn Berry seems to represent the direction in which the project is headed, as
+[stated by its lead
+maintainer](https://dev.to/arcanis/introducing-yarn-2-4eh1#conclusion). Given
+this, and the extent of the changes to the CLI, this RFC proposes that a `yarn-berry-install`
+buildpack be created to enable the new installation workflows.
+
+## Motivation
+
+As developers continue to adopt Yarn Berry, demand for the Node.js buildpack to
+support it has steadily increased.
+
+
+## Implementation
+
+
+### API
+
+`yarn-berry-install`
+
+Provides: `yarn_packages` OR `node_modules` Requires: `node`, during `build`
+
+The buildpack does not require `yarn` since the Berry release itself is meant
+to be checked into version control under `.yarn/releases`. Node v16.10 and
```

Comment:
Will we have to validate that `yarn` is vendored under `.yarn/releases` - either during `Detect` or during `Build`? Some of the workflows below invoke `yarn` so if users do not provide it under `.yarn/releases` and this buildpack does not require the `yarn` buildpack then those workflows will fail.

Perhaps more concretely, is it possible to pass `Detect` (by validating that `.yarnrc.yml` exists) and then fail at run time because `.yarn/releases` does not exist?

Reply:
The build would fail since `corepack` (which manages per-project installations of yarn & comes bundled with Node) won’t be able to locate the release as specified by `.yarnrc.yml`.  This raises additional questions and there will be challenges supporting offline builds with this approach. I’ll outline those issues in `Unresolved Questions`.